common/src/api/external/mod.rs

```diff
@@ -1832,11 +1833,122 @@ pub struct VpcFirewallRuleFilter {
 
 /// The protocols that may be specified in a firewall rule's filter
 #[derive(Clone, Copy, Debug, PartialEq, Deserialize, Serialize, JsonSchema)]
-#[serde(rename_all = "UPPERCASE")]
+#[serde(rename_all = "snake_case")]
+#[serde(tag = "type", content = "value")]
 pub enum VpcFirewallRuleProtocol {
     Tcp,
     Udp,
-    Icmp,
+    Icmp(Option<VpcFirewallIcmpFilter>),
+    // TODO: IPv6 not supported by instances.
+    // Icmpv6(Option<VpcFirewallIcmpFilter>),
+    // TODO: OPTE does not yet permit further L4 protocols. (opte#609)
+    // Other(u16),
+}
+
+impl FromStr for VpcFirewallRuleProtocol {
+    type Err = Error;
+
+    fn from_str(proto: &str) -> Result<Self, Self::Err> {
+        let (ty_str, content_str) = match proto.split_once(':') {
+            None => (proto, None),
+            Some((lhs, rhs)) => (lhs, Some(rhs)),
+        };
+
+        match (ty_str, content_str) {
+            (lhs, None) if lhs.eq_ignore_ascii_case("tcp") => Ok(Self::Tcp),
+            (lhs, None) if lhs.eq_ignore_ascii_case("udp") => Ok(Self::Udp),
+            (lhs, None) if lhs.eq_ignore_ascii_case("icmp") => {
+                Ok(Self::Icmp(None))
+            }
+            (lhs, Some(rhs)) if lhs.eq_ignore_ascii_case("icmp") => {
+                Ok(Self::Icmp(Some(rhs.parse()?)))
+            }
+            (lhs, None) => {
+                Err(Error::invalid_value(lhs, "unrecognized protocol"))
```

Comment:
I think elsewhere, when we use `invalid_value`, the idea is that the first part (the "label") is generally intended to be the _name_ or type of the field that was invalid, rather than the value. for example: https://github.com/oxidecomputer/omicron/blob/80e992229b52cecb517a59a0f686f8916549343f/nexus/src/app/switch_port.rs#L66-L72

so, I think a more consistent use of `invalid_value` would be:
```suggestion
                Err(Error::invalid_value(
                    "vpc_firewall_rule_protocol",
                    format!("unrecognized protocol: {lhs:?}"),
                ))
```

(and of course, this applies to other uses of `invalid_value` as well)